### ISSUE TITLE
HuBERT pre-training for streaming ASR, ngram LM biasing of streaming ASR

### ISSUE THREAD

**Issue by vesis84** [k2-fsa/icefall#1103]
Hello,
in a project, I am focusing on a streaming ASR (currently working with `pruned_transducer_stateless7`), 
and we were asked to use Self-supervised pre-training for it. 
And, I'd like to deploy the model in Sherpa, or sherpa-onnx afterwards.

I am thinking of taking the [zipformer encoder](https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless7/zipformer.py) from Librispeech recipe.
And add it as another encoder implementation for HuBERT training in ESPnet 
or some other toolkit (fairseq, torchaudio). 

But there might be some other possibility i am not aware of.
Would you have some suggestion?

Is the `pruned_transducer_stateless7` recipe the closest fit for streaming ASR with HuBERT pre-training?

Would it also be possible to use an ngram LM to bias the ASR system with in-domain text data?

Thank you,
Karel

**Comment by marcoyang1998** [comment]
> Is the pruned_transducer_stateless7 recipe the closest fit for streaming ASR with HuBERT pre-training?

`pruned_transducer_stateless7` is a non-streaming recipe, its streaming version is `pruned_transducer_stateless7_streaming`. We also publish the latest version of zipformer (see [here](https://github.com/k2-fsa/icefall/tree/master/egs/librispeech/ASR/zipformer)), which is currently our best streaming model.

Yes, you can use the encoder structure from the aforementioned recipes for HuBERT-style pre-training. Unfortunately, we don't have recipes for pre-training, you will need to implement it by yourself. I recommend to set up the pre-training pipeline in `icefall`, as this would make exporting the model easier.

> Would it also be possible to use an ngram LM to bias the ASR system with in-domain text data?

Yes, we have various decoding methods with ngram LM based on rescoring/shallow fusion. For transducer-based models, we have `fast_beam_search_*` and `modified_beam_search_ngram_rescoring`  link [here](https://github.com/k2-fsa/icefall/blob/7b0afbdc16066701759e088f7edbb648a0b879f0/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L2135). For CTC-based model, we have normal ngram rescoring methods, see [here](https://github.com/k2-fsa/icefall/blob/7b0afbdc16066701759e088f7edbb648a0b879f0/egs/librispeech/ASR/conformer_ctc2/decode.py#L41)

**Comment by danpovey** [comment]
If you want to make use of a Hubert-type model, one way to do it is to dump feature embeddings from Hubert and use them as an auxiliary loss.  I believe it was giving us an improvement, I'll ask Liyong to comment on where / details.
If there is plenty of data, I doubt pre-training would help much or make that much sense; I think it would be better to do self-supervised by getting a transcript.  Amounts to the same thing because of SpecAugment.

**Comment by glynpu** [comment]
> If you want to make use of a Hubert-type model, one way to do it is to dump feature embeddings from Hubert and use them as an auxiliary loss.

Currently, we dump code_book_indexes extracted from hubert embeddings.  https://github.com/k2-fsa/icefall/blob/03853f1ee5bc71b7ff5e36289682a430a196f75e/egs/librispeech/ASR/distillation_with_hubert.sh#L180

It's quite easy to modify it if you want to save hubert embeddings.

As for the WER improvements see Table 4 from [our mvq paper](https://arxiv.org/pdf/2211.00508.pdf)
<img width="610" alt="image" src="https://github.com/k2-fsa/icefall/assets/14951566/196b50b6-9e89-4928-8720-532c5ccba564">
